raiden/network/blockchain_service.py

```diff
@@ -146,31 +179,99 @@ def token(self, token_address: TokenAddress) -> Token:
         return self.address_to_token[token_address]
 
     def token_network_registry(self, address: TokenNetworkRegistryAddress) -> TokenNetworkRegistry:
-        if not is_binary_address(address):
-            raise ValueError("address must be a valid address")
 
         with self._token_network_registry_creation_lock:
             if address not in self.address_to_token_network_registry:
+
+                metadata = SmartContractMetadata(
+                    deployed_at=self.metadata.token_network_registry_deployed_at,
+                    address=Address(address),
+                    abi=self.contract_manager.get_contract_abi(CONTRACT_TOKEN_NETWORK_REGISTRY),
+                    runtime_bytecode=EVMBytecode(
+                        decode_hex(
+                            self.contract_manager.get_runtime_hexcode(
+                                CONTRACT_TOKEN_NETWORK_REGISTRY
+                            )
+                        )
+                    ),
+                    gas_measurements=gas_measurements(self.contract_manager.contracts_version),
+                    filters_start_at=self.metadata.filters_start_at,
+                )
+
                 self.address_to_token_network_registry[address] = TokenNetworkRegistry(
+                    jsonrpc_client=self.client, metadata=metadata, blockchain_service=self
+                )
+
+        return self.address_to_token_network_registry[address]
+
+    def token_network_from_registry(
+        self,
+        token_network_registry_address: TokenNetworkRegistryAddress,
+        token_address: TokenAddress,
+        block_identifier: BlockSpecification = "latest",
+    ) -> TokenNetwork:
+        token_network_registry = self.token_network_registry(token_network_registry_address)
+        token_network_address = token_network_registry.get_token_network(
+            token_address=token_address, block_identifier=block_identifier
+        )
+
+        if token_network_address is None:
+            raise ValueError(
+                f"{to_checksum_address(token_network_registry_address)} does not "
+                f"have the token {to_checksum_address(token_address)} "
+                f"registered."
+            )
+
+        with self._token_network_creation_lock:
+            if token_network_address not in self.address_to_token_network:
+                metadata = TokenNetworkMetadata(
+                    deployed_at=None,
```

Comment:
Questions:

1. Wouldn't this be invalid given the condition you have for raising a `ValueError` in case `deployed_at` and `filter_starts_at` are not the same?
2. If they're always the same, why not use a single value to keep track rather than having 2 values here.

Reply:
> Wouldn't this be invalid given the condition you have for raising a ValueError in case deployed_at and filter_starts_at are not the same?

Nope, `deployed_at` is ignored if it is `None`:

```python
        is_filter_start_valid = (
            self.deployed_at is not None and self.deployed_at == self.filters_start_at
        )
        if not is_filter_start_valid:
```

> If they're always the same, why not use a single value to keep track rather than having 2 values here.

The deployment block may be unknown,